### PR TITLE
G-API: oneVPL (simplification) Add simple decode pipeline

### PR DESCRIPTION
**Pull request by sivanov-work**
this PR is the single one in series of https://github.com/opencv/opencv/pull/20469
Introduces simple decoding pipeline & related data structures

P.S. This pipeline flow would be improved for fully async-decode flow in the next PRs

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### Build configuration

```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.3.0:20.04
build_image:Custom Win=openvino-2021.2.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+    mfxU32 half_width = width / 2;
+    mfxU32 half_height = height / 2;
+    switch (FourCC) {
+        case MFX_FOURCC_I420:
+        case MFX_FOURCC_NV12:
+            nbytes = width * height +  2 * half_width * half_height;
+            break;
+        case MFX_FOURCC_I010:
+        case MFX_FOURCC_P010:
+            nbytes = width * height + 2 * half_width * half_height;
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:37]: Should this size in bytes include alignment?

In some platforms, image data (stride) is aligned to some multiplier and may exceed its logical size.

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+                                  size_t out_buf_size)
+{
+    mfxU8* buf = reinterpret_cast<mfxU8*>(out_buf_ptr.get());
+    mfxU16 surfW = frameInfo.Width * 4;
+    mfxU16 surfH = frameInfo.Height;
+    (void)surfH;
+
+    // TODO more intelligent check
+    if (out_buf_size <= out_buf_ptr_offset) {
+        GAPI_LOG_WARNING(nullptr, "Not enough buffer, ptr: " << out_buf_ptr <<
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:63]: It seems it is not WARNING if then it is followed by `Assert`

Comment (dmatveev) [reply modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:63]: Aslo, maybe `"Insufficient buffer size"` or like is a more correct message?

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        [this] (EngineSession& sess) -> ExecutionStatus
+        {
+            return this->process_error(sess.last_status, static_cast<LegacyDecodeSession&>(sess));
+        }
+    );
+}
+
+void VPLLegacyDecodeEngine::initialize_session(mfxSession mfx_session,
+                                         DecoderParams&& decoder_param,
+                                         std::shared_ptr<onevpl::IDataProvider> provider)
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:173]: Alignment

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+                                                      provider);
+
+    sess_ptr->init_surface_pool(decode_pool_key);
+    // prepare working decode surface
+    sess_ptr->swap_surface(*this);
+}
+
+ProcessingEngineBase::ExecutionStatus VPLLegacyDecodeEngine::execute_op(operation_t& op, EngineSession& sess) {
+    return op(sess);
+}
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:224]: y do we need smt like this?

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        {
+            // The function requires more frame surface at output before decoding can proceed.
+            // This applies to external memory allocations and should not be expected for
+            // a simple internal allocation case like this
+            try {
+                sess.swap_surface(*this);
+                return ExecutionStatus::Continue;
+            } catch (const std::exception& ex) {
+                GAPI_LOG_WARNING(nullptr, "[" << sess.session << "] error: " << ex.what() <<
+                                          "Abort");
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:261]: in fact there's no abort. Also, is this handling required here?

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.hpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2021 Intel Corporation
+
+#ifndef GAPI_STREAMING_ONVPL_ENGINE_DECODE_DECODE_ENGINE_LEGACY_HPP
+#define GAPI_STREAMING_ONVPL_ENGINE_DECODE_DECODE_ENGINE_LEGACY_HPP
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.hpp:8]: engine/decode/decode_engine 🤯

File: modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ProcessingEngineBase::~ProcessingEngineBase() {
+    GAPI_LOG_INFO(nullptr, "destroyed");
+}
+
+ProcessingEngineBase::ExecutionStatus ProcessingEngineBase::process(mfxSession session) {
+    auto sess_it = sessions.find(session);
+    if (sess_it == sessions.end()) {
+
+        // TODO remember the last session status
+        return ExecutionStatus::Processed;
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp:30]: I don't get this logic, Why it is "processed"if it is not found?

File: modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    mfxU8 *p1 = bs.Data + bs.DataOffset;
+    if (bs.DataOffset > bs.MaxLength - 1) {
+        return MFX_ERR_NOT_ENOUGH_BUFFER;
+    }
+    if (bs.DataLength + bs.DataOffset > bs.MaxLength) {
+        return MFX_ERR_NOT_ENOUGH_BUFFER;
+    }
+    for (mfxU32 i = 0; i < bs.DataLength; i++) {
+        *(p0++) = *(p1++);
+    }
```
Comment (dmatveev) [inline modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp:118]: Is it copy or smt? may `copy_n` work here?

Comment (dmatveev) [reply modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp:118]: (or maybe some more optimized analogue)

**Review by dmatveev — COMMENTED** [review]
LGTM

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.hpp (hunk 6 above)
Comment (sivanov-work) [reply modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.hpp:8]: ok, i'll rename it..

File: modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp (hunk 7 above)
Comment (sivanov-work) [reply modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp:30]: i used it to notify `bool pull(Data&)` about no-more-data-for-processing and not-an-error situation with `false` 
Actually, maybe new status is required `NoData` or something like... Currently we have only single session from Source and it is predictable situation.
Will add new status here & update UTs

File: modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp (hunk 8 above)
Comment (sivanov-work) [reply modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp:118]: nice catch! actually it was stolen from oneVPL examples and looks like performance gap.
Interesting but Vtune doesn't point out this place in performance analysis

Comment (sivanov-work) [reply modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp:118]: will change on `std::copy_n` - let implementation to optimize it by itself

File: modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp (hunk 1 above)
Comment (dmatveev) [reply modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:39]: not sure, unfortunately. usually we learn about this limitations when enabling this or that particular platform. even more, on kmb one revision had one alignment and other revision has required the other.

At least I'd make it configurable somehow

File: modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp (hunk 7 above)
Comment (dmatveev) [reply modules/gapi/src/streaming/onevpl/engine/processing_engine_base.cpp]: ok

**Review by dmatveev — APPROVED** [review]
ok to merge

**Pull request merged**